### ISSUE TITLE
Release Hammer's source code

### ISSUE THREAD

**Issue by SamVanheer** [ValveSoftware/halflife#3740]
Release the source code for Hammer, the Half-Life 1 level editor. The last released version for this engine was [3.5 beta](http://web.archive.org/web/20040602212036/http://collective.valve-erc.com/index.php?go=hammer_beta) in April 2003, more than 20 years ago.

Most level designers nowadays use JACK, a closed source program with a paid version. A recent update broke compatibility with the last free version making it impossible to open JMF files saved using the paid version: https://steamcommunity.com/app/496450/discussions/0/4036980006268371571/

While the free version may receive updates to add support this illustrates a continuing problem with HL1 level design: there is no freely available level editor that works properly and has all of Hammer's features.

Hammer itself has had problems running on newer systems: AMD drivers have had problems for a long time and since it uses immediate mode OpenGL it has problems running well when loading very large maps. It also has bugs and non-functional user interface elements.

There are open source level editors but none that provide full support for Half-Life 1 and the only one in active development (Trenchbroom) is mainly a Quake level editor.

Providing the source code would allow an update to be developed to provide a direct replacement for Hammer that is open source and under the control of the community that uses it.

Some other things this would allow:
* A native Linux version
* Removing the use of Windows' [Short filename APIs](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#short-vs-long-names)
* Adding support for the engine's filesystem behavior (_downloads, _addon directories et al)
* Improving error reporting when a problem occurs during compilation (e.g. the very common "file not found" errors which don't report which file was not found)
* Ensuring the game is launched correctly to avoid problems when the working directory is not set to the game directory
* Fixing the spawnflags tab not properly setting flags after the 12th flag

**Comment by TarekLP** [comment]
This Would be really Cool if it was happeinng, but I believe TrenchBroom is supposed to be a replacement for it, since that is also Open Source

**Comment by SamVanheer** [comment]
Trenchbroom is not a replacement for Hammer. It uses a different workflow and does not implement all of Hammer's features. It's also mainly a Quake level editor, Half-Life support is largely provided by contributors, not maintainers.

I've made a proof-of-concept for a Qt-based UI: https://github.com/SamVanheer/Chisel-proof-of-concept

This is a mockup that shows what a modernized UI could look like. With Hammer's source code in hand we could overhaul the program to work well on modern versions of Windows as well as Linux and deal with graphics performance and compatibility issues. 

On top of that new features can be added as needed without having to rely on closed source projects or hoping contributions to level editors meant for other game engines are merged in.

**Comment by Aynekko** [comment]
> A recent update broke compatibility with the last free version making it impossible to open JMF files saved using the paid version:

Free version of Jack was updated, it's now up to date with the paid one.

**Comment by SamVanheer** [comment]
That's great! Still doesn't change anything though.

**Comment by Aynekko** [comment]
> That's great! Still doesn't change anything though.

Well yes, it's not like I disagree, the hammer code would be neat, because Jack's owner doesn't agree to implement all things suggested and it's closed-source... Just corrected the fact in your original post.

**Comment by karakakarkarkar** [comment]
Hammertime editor (https://github.com/Duude92/hammertime/tree/development) is not fully replacement for Hammer 3.5 beta and Jackhammer?

**Comment by karakakarkarkar** [comment]
@SamVanheer leaked hl2 sources not contained fully worldcraft (valve hammer editor) source code?

**Comment by RenPT23R** [comment]
OPEN THE FUCKING SOURCE CODE!!!!!!!!!!!!!

**Comment by stevenlafl** [comment]
@kisak-valve Would you mind asking colleagues to provide an official stance on this? It would really be great to fork and community maintain it, as Valve typically doesn't take open source contributions (I've created PRs for Proton but they end up being reimplemented later)

**Comment by sabianroberts** [comment]
I agree, I think open-sourcing Hammer would be massively beneficial to all Half-Life 1 and Source developers.

**Comment by Rabtos** [comment]
💯🔥🔥🐦‍🔥🔥🔥
